lib/membrane_video_compositor_multiple_inputs.ex

```diff
@@ -0,0 +1,294 @@
+    end
+  end
+
+  defp push_frame(ids_to_tracks, id, frame) do
+    Map.update!(ids_to_tracks, id, fn track ->
+      Track.push_frame(track, frame)
+    end)
+  end
+
+  defp all_have_frame?(ids_to_tracks) when map_size(ids_to_tracks) == 0 do
+    false
+  end
+
+  defp all_have_frame?(ids_to_tracks) do
+    ids_to_tracks
+    |> Map.values()
+    |> Enum.all?(&Track.has_frame?/1)
+  end
+
+  defp get_ids_to_frames(ids_to_tracks) do
+    ids_to_tracks
+    |> Enum.map(fn {id, %Track{} = track} -> {id, Track.first_frame(track)} end)
+    |> Map.new()
+  end
+
+  defp pop_frames(ids_to_tracks) do
+    ids_to_tracks
+    |> Enum.map(fn {id, %Track{} = track} ->
+      {id, Track.pop_frame(track)}
+    end)
+    |> Map.new()
+  end
+
+  defp remove_depleted_tracks(%{ids_to_tracks: ids_to_tracks} = state) do
+    ids_to_tracks
+    |> Enum.reduce(
+      state,
+      fn {id, %Track{} = track}, state ->
+        if Track.depleted?(track) do
+          remove_track(state, id)
+        else
+          state
+        end
+      end
+    )
+  end
+
+  defp track_depleted?(ids_to_tracks, id) do
+    track = Map.get(ids_to_tracks, id)
+    Track.depleted?(track)
+  end
+
+  defp remove_track(%{ids_to_tracks: ids_to_tracks, internal_state: internal_state} = state, id) do
+    ids_to_tracks = Map.delete(ids_to_tracks, id)
+    {:ok, internal_state} = state.compositor_module.remove_video(id, internal_state)
+    %{state | ids_to_tracks: ids_to_tracks, internal_state: internal_state}
+  end
+
+  defp update_track_status(ids_to_tracks, id, status) do
+    ids_to_tracks
+    |> Map.update!(id, fn track -> %Track{track | status: status} end)
+  end
+
+  defp tracks_status(ids_to_tracks) when map_size(ids_to_tracks) == 0 do
+    :all_finished
+  end
+
+  defp tracks_status(ids_to_tracks) do
+    if Map.values(ids_to_tracks) |> Enum.all?(&Track.depleted?/1) do
+      :all_finished
+    else
+      :still_ongoing
+    end
+  end
+
+  @impl true
+  def handle_end_of_stream(
+        pad,
+        _context,
+        %{ids_to_tracks: ids_to_tracks, pads_to_ids: {_new_id, pads_to_ids}} = state
+      ) do
+    id = Map.get(pads_to_ids, pad)
+
+    ids_to_tracks = ids_to_tracks |> update_track_status(id, :end_of_stream)
+    state = %{state | ids_to_tracks: ids_to_tracks}
+
+    state =
+      if track_depleted?(ids_to_tracks, id) do
+        remove_track(state, id)
+      else
+        state
+      end
+
+    {{buffers_status, buffers}, state} = merge_frames(state)
+
+    case {buffers_status, tracks_status(state.ids_to_tracks)} do
+      {:empty, :still_ongoing} ->
+        {:ok, state}
+
+      {:empty, :all_finished} ->
+        {{:ok, end_of_stream: :output}, state}
+
+      {:merged, :still_ongoing} ->
+        {{:ok, buffer: {:output, buffers}}, state}
+
+      {:merged, :all_finished} ->
+        {{:ok, buffer: {:output, buffers}, end_of_stream: :output}, state}
+    end
+  end
+
+  @dialyzer {:nowarn_function, determine_compositor_module: 1}
```

Comment:
maybe `Implementation.get_implementation_module` could be excluded instead?

Reply:
It is a temporary solution. I leave it as it is.